packages/components/table/src/store/watcher.ts

```diff
@@ -213,24 +217,37 @@ function useWatcher<T>() {
     isAllSelected.value = value
 
     let selectionChanged = false
-    let childrenCount = 0
-    const rowKey = instance?.store?.states?.rowKey.value
-    data.value.forEach((row, index) => {
-      const rowIndex = index + childrenCount
-      if (selectable.value) {
-        if (
-          selectable.value.call(null, row, rowIndex) &&
-          toggleRowStatus(selection.value, row, value)
-        ) {
-          selectionChanged = true
+    let rowIndex = 0
+    const { childrenColumnName } = instance.store.states
+    const treeProps = {
+      children: childrenColumnName.value,
+      checkStrictly: false, // Disable checkStrictly when selecting all
+    }
+
+    const toggleSelectedStatus = (data: DefaultRow[]) => {
+      data.forEach((row) => {
+        if (selectable.value) {
+          if (
+            selectable.value.call(null, row, rowIndex) &&
```

Comment:
Here the `rowIndex` variable can be removed, and the subscript of the forEach loop can be used directly.